[WebGPU] Implement GPUExternalTexture

mwyrzykowski opened the pull request:
#### cb91b749f30d1cc1bb01bfce9adbe18ad3cea698
<pre>
[WebGPU] Implement GPUExternalTexture
<a href="https://bugs.webkit.org/show_bug.cgi?id=255883">https://bugs.webkit.org/show_bug.cgi?id=255883</a>
&lt;radar://108463707&gt;

Reviewed by Myles C. Maxfield.

Implement GPUExternalTexture and pass it through to WebGPU.framework.

* Source/WebCore/Modules/WebGPU/GPUExternalTextureDescriptor.h:
(WebCore::GPUExternalTextureDescriptor::convertToBacking const):

* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUConvertToBackingContext.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
(PAL::WebGPU::DeviceImpl::importExternalTexture):
(PAL::WebGPU::DeviceImpl::createBindGroupLayout):
(PAL::WebGPU::DeviceImpl::createBindGroup):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDowncastConvertToBackingContext.cpp:
(PAL::WebGPU::DowncastConvertToBackingContext::convertToBacking):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDowncastConvertToBackingContext.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUExternalTextureImpl.cpp:
(PAL::WebGPU::ExternalTextureImpl::ExternalTextureImpl):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUExternalTextureImpl.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUExternalTextureDescriptor.h:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/APIConversions.h:
(WebGPU::fromAPI):
* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::Device::createBindGroup):
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::Device):
(WebGPU::metalPixelFormatForCVPixelBuffer):
(WebGPU::Device::createTextureFromPixelBuffer const):
(wgpuDeviceCreateExternalTexture):
* Source/WebGPU/WebGPU/ExternalTexture.h: Added.
(WebGPU::ExternalTexture::create):
(WebGPU::ExternalTexture::createInvalid):
(WebGPU::ExternalTexture::pixelBuffer const):
* Source/WebGPU/WebGPU/ExternalTexture.mm: Added.
(WebGPU::Device::createExternalTexture):
(WebGPU::ExternalTexture::ExternalTexture):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::createRemoteGPU):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteAdapter.cpp:
(WebKit::RemoteAdapter::RemoteAdapter):
(WebKit::RemoteAdapter::requestDevice):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteAdapter.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
(WebKit::RemoteDevice::RemoteDevice):
(WebKit::RemoteDevice::importExternalTexture):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
(WebKit::RemoteGPU::RemoteGPU):
(WebKit::RemoteGPU::remoteRenderingBackend const):
(WebKit::RemoteGPU::requestAdapter):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderPassEncoder.cpp:
(WebKit::RemoteRenderPassEncoder::setBindGroup):
* Source/WebKit/Shared/WebGPU/WebGPUExternalTextureDescriptor.cpp:
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/WebGPU/WebGPUExternalTextureDescriptor.h:
* Source/WebKit/Shared/WebGPU/WebGPUExternalTextureDescriptor.serialization.in:
* Websites/webkit.org/demos/webgpu/external-textured-cube.html: Added.
* Websites/webkit.org/demos/webgpu/scripts/external-textured-cube.js: Added.
(async helloCube.frameUpdate):
(async helloCube):

Canonical link: <a href="https://commits.webkit.org/264108@main">https://commits.webkit.org/264108@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/08cf6a72556d3b75a285cdbcd123f13014488db9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6726 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6943 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7127 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8315 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/6985 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6724 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8076 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6888 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/9886 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6844 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7407 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6160 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8409 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/4852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6086 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/13906 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/5671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/6615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6164 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8905 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/6307 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6657 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5451 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/6865 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6046 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1605 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1585 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10221 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/7049 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6419 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1738 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
```diff
@@ -151,9 +157,26 @@ void RemoteDevice::importExternalTexture(const WebGPU::ExternalTextureDescriptor
+    if (!convertedDescriptor || !convertedDescriptor->m_mediaIdentifier.has_value())
         return;
 
+    BinarySemaphore semaphore;
```
The BinarySemaphore is not needed here, since `performWithMediaPlayerOnMainThread ` already uses one via `callOnMainAndWait`

I will remove it

litherum commented on Source/WebGPU/WebGPU/MetalSPI.h:
```diff
@@ -0,0 +1,88 @@
+#if USE(INTERNAL_APPLE_SDK)
+#import <Metal/MTLPixelFormat_Private.h>
+#else
+typedef NS_ENUM(NSUInteger, WebKitMTLPixelFormatSPI) {
```
What would it take to implement this without any SPI?

I think we should try very hard not to use any SPI. I think we can implement external texture by just using multiple MTLTextures to represent multiplayer formats, instead of these SPIs.

mwyrzykowski replied:
I don't think its possible to implement external textures without SPIs without substantial performance regression from the SPI texture format route which is already implemented by the CoreVideo.framework

litherum commented on Source/WebGPU/WebGPU/WebGPUExt.h:
```diff
@@ -177,6 +186,7 @@ WGPU_EXPORT void wgpuAdapterRequestDeviceWithBlock(WGPUAdapter adapter, WGPUDevi
 WGPU_EXPORT void wgpuBufferMapAsyncWithBlock(WGPUBuffer buffer, WGPUMapModeFlags mode, size_t offset, size_t size, WGPUBufferMapBlockCallback callback);
 WGPU_EXPORT void wgpuDeviceCreateComputePipelineAsyncWithBlock(WGPUDevice device, WGPUComputePipelineDescriptor const * descriptor, WGPUCreateComputePipelineAsyncBlockCallback callback);
 WGPU_EXPORT void wgpuDeviceCreateRenderPipelineAsyncWithBlock(WGPUDevice device, WGPURenderPipelineDescriptor const * descriptor, WGPUCreateRenderPipelineAsyncBlockCallback callback);
+WGPU_EXPORT WGPUExternalTexture wgpuDeviceCreateExternalTexture(WGPUDevice device, const WGPUExternalTextureDescriptor* descriptor);
```
The JS API calls this "import", not "create"

litherum commented on Source/WebGPU/WebGPU/WebGPUExt.h:
```diff
@@ -71,6 +72,14 @@ typedef struct WGPUSurfaceDescriptorCocoaCustomSurface {
+typedef struct WGPUExternalTextureDescriptor {
+    WGPUChainedStruct const * nextInChain;
+    char const * label; // nullable
+    void* mediaIdentifier; // CVPixelBufferRef
```
WebGPUExt.h is just for Apple platforms, so we can totally use CoreVideo types in here. No need for `void*`

Can this single `CVPixelBufferRef` handle both `HTMLVideoElement`s as well as `VideoFrame`s?

How are we handling the `colorspace` and `label` fields of `GPUExternalTextureDescriptor`?

mwyrzykowski replied:
Yes it can support `HTMLVideoElement` and `VideoFrame`. For `HTMLVideoElement`, I just get the `VideoFrame` at the current time. So for `VideoFrame`, we can just use it directly.

In both cases the `CVPixelBuffer` comes from the `VideoFrame`

mwyrzykowski replied:
Regarding colorspace, I don’t use that currently. It should be used when selecting the appropriate MTLPixelFormat

litherum commented on Source/WebGPU/WebGPU/WebGPUExt.h:
```diff
@@ -71,6 +72,14 @@ typedef struct WGPUSurfaceDescriptorCocoaCustomSurface {
     WGPUCompositorIntegrationRegisterBlockCallback compositorIntegrationRegister;
 } WGPUSurfaceDescriptorCocoaCustomSurface;
 
+typedef struct WGPUExternalTextureImpl* WGPUExternalTexture;
```
Shouldn't this be defined higher in the file, to keep the same organization as WebGPU.h?

mwyrzykowski replied:
Good idea, I will move it up

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -88,13 +88,19 @@ static MTLRenderStages metalRenderStage(ShaderStage shaderStage)
+            if (entry.nextInChain->sType != static_cast<WGPUSType>(WGPUSTypeExtended_BindGroupEntryExternalTexture))
+                return BindGroup::createInvalid(*this);
+
+            externalTexture = const_cast<WGPUExternalTexture>(reinterpret_cast<const WGPUExternalTextureImpl*>(entry.nextInChain->next));
```
I think we should wrap the external texture in a struct anyway, even if it's the only field, because there's gotta be more stuff that we'll eventually want to add next to the external texture.

mwyrzykowski replied:
Its not clear to me we will need more than `WGPUExternalTexture`, I would propose making that change when we need to add items for external texture support

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -88,13 +88,19 @@ static MTLRenderStages metalRenderStage(ShaderStage shaderStage)
         bool samplerIsPresent = WebGPU::samplerIsPresent(entry);
         bool textureViewIsPresent = WebGPU::textureViewIsPresent(entry);
-        if (bufferIsPresent + samplerIsPresent + textureViewIsPresent != 1)
+        bool externalTextureIsPresent = !!externalTexture;
```
please feel free to disagree, but I cringe every time I see `!!` instead of `static_cast<bool>()`

mwyrzykowski replied:
It seems to be more common in WebKit to use `!!` than `static_cast<bool>()` so I will leave it, but I didn't analyze all the occurrences Find brought up

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -88,13 +88,19 @@ static MTLRenderStages metalRenderStage(ShaderStage shaderStage)
         bool textureViewIsPresent = WebGPU::textureViewIsPresent(entry);
-        if (bufferIsPresent + samplerIsPresent + textureViewIsPresent != 1)
+        bool externalTextureIsPresent = !!externalTexture;
+        if (bufferIsPresent + samplerIsPresent + textureViewIsPresent + externalTextureIsPresent != 1)
```
I realize you didn't write this, but `!= 1` (whitespace)

mwyrzykowski replied:
I don't see a whitespace error and the style checker doesn't either.

litherum replied:
Yep, you're right

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -116,6 +122,13 @@ static MTLRenderStages metalRenderStage(ShaderStage shaderStage)
+            } else if (externalTextureIsPresent) {
+                auto pixelBuffer = WebGPU::fromAPI(externalTexture).pixelBuffer();
+                id<MTLTexture> texture = createTextureFromPixelBuffer(pixelBuffer);
+                [argumentEncoder[stage] setTexture:texture atIndex:index++];
```
I don't understand. The whole point of external textures is that they are "fat" objects, that include way more than just a texture. I expected to see a shading language component, where sample operations on external textures get codegen'ed to a big block of code.

Even if that isn't all implemented in this first patch, I think we should still think about where/how to handle "fat" textures, and leave a space here to be filled-in with the relevant code that will come in a follow-up patch.

mwyrzykowski replied:
By using the MTLPixelFormat SPI we avoid need for this code-gen, it makes it simpler and more performant

litherum commented on Source/WebGPU/WebGPU/Device.h:
```diff
@@ -58,6 +60,8 @@ class Sampler;
 class ShaderModule;
 class Texture;
 
+typedef struct __CVBuffer *CVPixelBufferRef;
```
We can totally just include `<CoreVideo/CoreVideo.h>`

litherum commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -151,6 +153,9 @@ static void captureFrame(id<MTLDevice> captureObject)
 #endif
 #endif
 
+#if COREVIDEO_SUPPORTS_METAL
```
When would this ever be false? Doesn't Core Video always support Metal on all our platforms?

mwyrzykowski replied:
watchOS

litherum replied:
CoreVideo is on watchOS (according to https://developer.apple.com/documentation/corevideo?language=objc) and we build watchOS with a version of Metal.framework custom declared: https://github.com/WebKit/WebKit/blob/main/Tools/Scripts/configure-xcode-for-embedded-development#L327

Is it possible to augment `configure-xcode-for-embedded-development` to make it so that we don't have to deal with both `COREVIDEO_SUPPORTS_METAL` and `!COREVIDEO_SUPPORTS_METAL`, and we can just unconditionally use this code on all platforms?

litherum replied:
(discussed offline, resolving conversation)

litherum reviewed (commented):
Not finished reviewing yet, but I have to go to a meeting now

litherum commented on Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
```diff
@@ -146,9 +148,27 @@ Ref<Sampler> DeviceImpl::createSampler(const SamplerDescriptor& descriptor)
 }
 
-Ref<ExternalTexture> DeviceImpl::importExternalTexture(const ExternalTextureDescriptor&)
+static CFStringRef convertToCGColorSpace(const PredefinedColorSpace& colorSpace)
```
Nit: This doesn't return a CGColorSpace, contrary to the name of the function

mwyrzykowski replied:
It returns the name of a CGColorSpace, do you think `convertToCGColorSpaceName` is better?

litherum replied:
Yes, I think that's a better name.

From my comment below, I think this function should be moved to inside WebGPU.framework, and the API would just have a 2-value enum.

litherum commented on Source/WebGPU/WebGPU/Device.h:
```diff
@@ -76,6 +80,7 @@ class Device : public WGPUDeviceImpl, public ThreadSafeRefCountedAndCanMakeThrea
     Ref<CommandEncoder> createCommandEncoder(const WGPUCommandEncoderDescriptor&);
     Ref<ComputePipeline> createComputePipeline(const WGPUComputePipelineDescriptor&);
     void createComputePipelineAsync(const WGPUComputePipelineDescriptor&, CompletionHandler<void(WGPUCreatePipelineAsyncStatus, Ref<ComputePipeline>&&, String&& message)>&& callback);
+    Ref<ExternalTexture> createExternalTexture(const WGPUExternalTextureDescriptor&);
```
the JS spec calls this "import" not "create"

mwyrzykowski replied:
That is true however it creates an `ExternalTexture`, so I think we should deviate from the JS specification since it doesn't align with the internal implementation in this case.

litherum commented on Source/WebGPU/WebGPU/WebGPUExt.h:
```diff
@@ -177,6 +193,7 @@ WGPU_EXPORT void wgpuAdapterRequestDeviceWithBlock(WGPUAdapter adapter, WGPUDevi
 WGPU_EXPORT void wgpuBufferMapAsyncWithBlock(WGPUBuffer buffer, WGPUMapModeFlags mode, size_t offset, size_t size, WGPUBufferMapBlockCallback callback);
 WGPU_EXPORT void wgpuDeviceCreateComputePipelineAsyncWithBlock(WGPUDevice device, WGPUComputePipelineDescriptor const * descriptor, WGPUCreateComputePipelineAsyncBlockCallback callback);
 WGPU_EXPORT void wgpuDeviceCreateRenderPipelineAsyncWithBlock(WGPUDevice device, WGPURenderPipelineDescriptor const * descriptor, WGPUCreateRenderPipelineAsyncBlockCallback callback);
+WGPU_EXPORT WGPUExternalTexture wgpuDeviceImportExternalTexture(WGPUDevice device, const WGPUExternalTextureDescriptor* descriptor);
```
I think this is the right design (putting this here)

litherum commented on Source/WebGPU/WebGPU/WebGPUExt.h:
```diff
@@ -71,6 +80,13 @@ typedef struct WGPUSurfaceDescriptorCocoaCustomSurface {
+typedef struct WGPUExternalTextureDescriptor {
+    WGPUChainedStruct const * nextInChain;
+    char const * label; // nullable
+    CVPixelBufferRef pixelBuffer;
```
ooooh I like it

litherum commented on Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
```diff
@@ -213,9 +233,15 @@ Ref<BindGroup> DeviceImpl::createBindGroup(const BindGroupDescriptor& descriptor
+    auto backingEntries = descriptor.entries.map([this, &chainedEntries] (const auto& bindGroupEntry) {
+        auto externalTexture = std::holds_alternative<std::reference_wrapper<ExternalTexture>>(bindGroupEntry.resource) ? m_convertToBackingContext->convertToBacking(std::get<std::reference_wrapper<ExternalTexture>>(bindGroupEntry.resource).get()) : nullptr;
+        chainedEntries.append(WGPUChainedStruct {
+            .next = reinterpret_cast<WGPUChainedStruct*>(externalTexture),
```
I don't think this is right. `externalTexture` is an object, not a descriptor. These "chain" things are so we can add new functionality without breaking backwards compatibility; they're supposed to be an open-ended linked list of objects to chain together. I think this should be

```
// in WebGPUExt.h
struct WGPUBindGroupExternalTextureEntry {
    WGPUChainedStruct chain;
    WGPUExternalTexture externalTexture;
};

// here
WGPUBindGroupExternalTextureEntry externalTextureEntry;
auto externalTexture = ... holds_alternative<>() ? ... : nullptr;
externalTextureEntry.chain.next = nullptr;
externalTextureEntry.chain.sType = WGPUSTypeExtended_BindGroupEntryExternalTexture;
externalTextureEntry.externalTexture = externalTexture;
return WGPUBindGroupEntry {
    &externalTextureEntry.chain,
    ...
};
```

mwyrzykowski replied:
I can certainly do that, it's not necessary right now, but maybe it is less confusing so it seems reasonable to make that change

litherum commented on Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUExternalTextureDescriptor.h:
```diff
@@ -31,7 +31,9 @@
 struct ExternalTextureDescriptor : public ObjectDescriptorBase {
+    uint64_t mediaIdentifier { 0 };
     PredefinedColorSpace colorSpace { PredefinedColorSpace::SRGB };
+    void *pixelBuffer { nullptr };
```
Why can't we use the real CV type here? I think we can. CV is part of the platform.

mwyrzykowski replied:
That's true but we are in the platform abstraction layer here. So we can `typedef struct __CVBuffer* CVPixelBufferRef;` but that seems odd adding that inside PAL.

litherum replied:
hrmmm I see. ok

litherum replied:
Should we rename this to WebGPUExternalTextureDescriptorCocoa.h? Then we could use real types.

mwyrzykowski replied:
I think `typedef struct __CVBuffer* CVPixelBufferRef;` is the best compromise in terms of type safety as adding `WebGPUExternalTextureDescriptorCocoa.h` is not needed right now given the single implementation of WebGPU we currently have. When we have multiple implementations we can do something about `typedef struct __CVBuffer* CVPixelBufferRef;` inside PAL

litherum commented on Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUExternalTextureDescriptor.h:
```diff
@@ -31,7 +31,9 @@
 namespace PAL::WebGPU {
 
 struct ExternalTextureDescriptor : public ObjectDescriptorBase {
+    uint64_t mediaIdentifier { 0 };
```
How is this going to work in WebKit1?

mwyrzykowski replied:
No plans to support WebGPU in WebKit1

litherum replied:
webgpu already works in wk1! would be sad if we couldn't keep it working

litherum commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.h:
```diff
@@ -103,6 +107,7 @@ class RemoteGPU final : public IPC::StreamMessageReceiver {
     RefPtr<IPC::StreamServerConnection> m_streamConnection;
     RefPtr<PAL::WebGPU::GPU> m_backing WTF_GUARDED_BY_CAPABILITY(workQueue());
     Ref<WebGPU::ObjectHeap> m_objectHeap WTF_GUARDED_BY_CAPABILITY(workQueue());
+    Ref<RemoteRenderingBackend> m_remoteRenderingBackend;
```
Isn't this a retain cycle? I think we should retain the resource heap, not the RRB.

mwyrzykowski replied:
We need the RRB to call `performWithMediaPlayerOnMainThread`:
```
remoteRenderingBackend().performWithMediaPlayerOnMainThread
```

Alternatively we can re-implement `performWithMediaPlayerOnMainThread`:

```
    callOnMainRunLoopAndWait([&, gpuConnectionToWebProcess = m_gpuConnectionToWebProcess, identifier] {
        if (auto player = gpuConnectionToWebProcess->remoteMediaPlayerManagerProxy().mediaPlayer(identifier))
            callback(*player);
    });
```

mwyrzykowski replied:
I will break the retain cycle, thank you for noticing

litherum commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.h:
```diff
@@ -70,10 +72,12 @@ class RemoteGPU final : public IPC::StreamMessageReceiver {
 
     void stopListeningForIPC();
 
+    RemoteRenderingBackend& remoteRenderingBackend() const;
```
I think this is a layering violation. Can we do some sort encapsulation to avoid this? Either have some kind of delegation pattern, or use a WTF::Function or something?

Can we do it in such a way where the RemoteGPU doesn't have to retain anything, but the RemoteDevice does? (Because RemoteDevice is the only place that actually uses this thing)

mwyrzykowski replied:
That should be possible. Currently it is used like so:
```
m_remoteGPU->remoteRenderingBackend().performWithMediaPlayerOnMainThread(WebCore::MediaPlayerIdentifier(convertedDescriptor->mediaIdentifier), [&convertedDescriptor](auto& player) mutable {
```
we should be able to translate that into:
```
m_performWithMediaPlayerOnMainThread(convertedDescriptor->mediaIdentifier), [&convertedDescriptor](auto& player) mutable {
```

I will attempt that refactor

litherum commented on Source/WebGPU/WebGPU/WebGPUExt.h:
```diff
@@ -71,6 +80,13 @@ typedef struct WGPUSurfaceDescriptorCocoaCustomSurface {
+    WGPUChainedStruct const * nextInChain;
+    char const * label; // nullable
+    CVPixelBufferRef pixelBuffer;
+    CFStringRef colorSpaceName;
```
Is there a reason you're doing this instead of making a new 2-value enum here? That seems cleaner.

mwyrzykowski replied:
Sure I will add an enum

litherum commented on Source/WebKit/Shared/WebGPU/WebGPUExternalTextureDescriptor.h:
```diff
@@ -34,6 +34,7 @@
 namespace WebKit::WebGPU {
 
 struct ExternalTextureDescriptor : public ObjectDescriptorBase {
+    uint64_t mediaIdentifier { 0 };
```
Can we use WebCore::MediaPlayerIdentifier instead?

mwyrzykowski replied:
I think so because we are not inside PAL in this case, I will try that

litherum commented on Source/WebKit/Shared/WebGPU/WebGPUExternalTextureDescriptor.serialization.in:
```diff
@@ -22,6 +22,7 @@
 
 #if ENABLE(GPU_PROCESS)
 [AdditionalEncoder=StreamConnectionEncoder] struct WebKit::WebGPU::ExternalTextureDescriptor : WebKit::WebGPU::ObjectDescriptorBase {
+    uint64_t mediaIdentifier;
```
ditto

litherum commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
```diff
@@ -151,9 +156,21 @@ void RemoteDevice::importExternalTexture(const WebGPU::ExternalTextureDescriptor
+    if (!convertedDescriptor || !convertedDescriptor->mediaIdentifier)
         return;
 
+    m_remoteGPU->remoteRenderingBackend().performWithMediaPlayerOnMainThread(WebCore::MediaPlayerIdentifier(convertedDescriptor->mediaIdentifier), [&convertedDescriptor](auto& player) mutable {
```
Is this synchronous? We need to make sure it's synchronous

mwyrzykowski replied:
It is :)

```
void RemoteRenderingBackend::performWithMediaPlayerOnMainThread(MediaPlayerIdentifier identifier, Function<void(MediaPlayer&)>&& callback)
{
    callOnMainRunLoopAndWait
```

litherum reviewed (commented):
90% done reviewing, but have to leave for (another) meeting :(

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -116,6 +122,25 @@ static MTLRenderStages metalRenderStage(ShaderStage shaderStage)
+            } else if (externalTextureIsPresent) {
+                auto& externalTexture = WebGPU::fromAPI(wgpuExternalTexture);
+                auto textureData = createExternalTextureFromPixelBuffer(externalTexture.pixelBuffer(), externalTexture.colorSpaceName());
+                [argumentEncoder[stage] setTexture:textureData.texture0 atIndex:index++];
+                [argumentEncoder[stage] setTexture:textureData.texture1 atIndex:index++];
+                ASSERT(textureData.texture0);
+                ASSERT(textureData.texture1);
+                if (textureData.texture0)
+                    resources.append({ textureData.texture0, MTLResourceUsageRead, metalRenderStage(stage) });
+                if (textureData.texture1)
+                    resources.append({ textureData.texture1, MTLResourceUsageRead, metalRenderStage(stage) });
```
Based on internal discussions, it sounds like this code is changing soon?

mwyrzykowski replied:
I don't recall, why would this be removed?

litherum replied:
not removed, but changed - to use a central place that knows how to pack/rearrange the fields optimally

mwyrzykowski replied:
Oh yes, we should do something like that

litherum commented on Source/WebGPU/WebGPU/BindGroupLayout.mm:
```diff
@@ -42,10 +42,19 @@
     auto descriptor = [MTLArgumentDescriptor new];
-    descriptor.dataType = MTLDataTypePointer;
+    auto bufferType = buffer.type;
+    if (bufferType == static_cast<uint32_t>(WGPUBufferBindingType_Float3x3)) {
```
I don't understand this. `WGPUBufferBindingType` determines whether the buffer is a uniform buffer or storage buffer, not whether it holds a float3x3 or a float4x3. Float3x3s are not "bound"; whole buffers are bound.

mwyrzykowski replied:
For external textures, we would bind the float3x3 directly using `*[argumentEncoder[stage] constantDataAtIndex:index++] = variable_float3x3matrix;` for instance

litherum commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -360,6 +364,54 @@ static void captureFrame(id<MTLDevice> captureObject)
+    simd::float3x2 uvRemappingMatrix(isFlipped ? simd::make_float2(1.f, -1.f) : simd::make_float2(1.f, 1.f));
+    simd::float4x3 colorSpaceConversionMatrix = colorSpaceConversionMatrixForPixelBuffer(pixelBuffer);
+
+    return { mtlTexture0, mtlTexture1, uvRemappingMatrix, colorSpaceConversionMatrix };
```
In my benchmark testing, I found that simply retaining the MTLTexture wasn't enough; I was getting intermittent crashes doing that. I had to retain the CVMetalTextures themselves (and then the crashes went away). There's a note in [the docs](https://developer.apple.com/documentation/corevideo/1456754-cvmetaltexturecachecreatetexture?language=objc) which is difficult to understand, but I think (probably?) describes this

mwyrzykowski replied:
Oh thanks for noticing, looks like we should not release until the completion handler of the command buffer.

litherum commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -360,6 +364,54 @@ static void captureFrame(id<MTLDevice> captureObject)
+        isFlipped = CVMetalTextureIsFlipped(plane0);
+        ASSERT(isFlipped == CVMetalTextureIsFlipped(plane1));
+        CFRelease(plane0);
+        CFRelease(plane1);
```
Are we guaranteeing that both `plane0` and `plane1` are both non-null?

mwyrzykowski replied:
oh good point, I will break this up

litherum commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -360,6 +364,54 @@ static void captureFrame(id<MTLDevice> captureObject)
+        CFRelease(plane1);
+    }
+
+    simd::float3x2 uvRemappingMatrix(isFlipped ? simd::make_float2(1.f, -1.f) : simd::make_float2(1.f, 1.f));
```
seems confusing to create a float3x2 with just a float2

mwyrzykowski replied:
the constructor fills the diagonal

mwyrzykowski replied:
but I realized I had my matrix for y-flip wrong, it should be:
```
[1   0   0]
[0 -1   1]
```
this is corrected now

litherum commented on Source/WebGPU/WebGPU/Device.h:
```diff
@@ -130,6 +135,13 @@ class Device : public WGPUDeviceImpl, public ThreadSafeRefCountedAndCanMakeThrea
+        simd::float3x2 uvRemappingMatrix;
+        simd::float4x3 colorSpaceConversionMatrix;
+    };
+    ExternalTextureData createExternalTextureFromPixelBuffer(CVPixelBufferRef, CFStringRef) const;
```
Should this be here? Seems like a better fit for `BindGroup` or `BindGroupLayout`, since it's only called from there.

Maybe not, because it uses the CVMetalTextureCache, and the cache should definitely be owned by the device (or adapter).

Maybe this should be a static free function in BindGroup.mm, taking the cache as a parameter?

mwyrzykowski replied:
I can move both `Device::createExternalTextureFromPixelBuffer` and `colorSpaceConversionMatrixForPixelBuffer` to BindGroup.mm

litherum commented on Source/WebGPU/WebGPU/Device.h:
```diff
@@ -153,6 +165,9 @@ class Device : public WGPUDeviceImpl, public ThreadSafeRefCountedAndCanMakeThrea
     HardwareCapabilities m_capabilities { };
 
     const Ref<Adapter> m_adapter;
+#if defined(__OBJC__) && COREVIDEO_SUPPORTS_METAL
```
`CVMetalTextureCacheRef` is a CF type, so I don't think there's a need for `defined(__OBJC__)`.

Also, the top of this file already uses `#import` instead of `#include` so this file is already only valid from obj-c anyway.

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -116,6 +122,25 @@ static MTLRenderStages metalRenderStage(ShaderStage shaderStage)
+                    resources.append({ textureData.texture0, MTLResourceUsageRead, metalRenderStage(stage) });
+                if (textureData.texture1)
+                    resources.append({ textureData.texture1, MTLResourceUsageRead, metalRenderStage(stage) });
+                // FIXME: encode CSC and UV remapping after https://bugs.webkit.org/show_bug.cgi?id=256721 is implemented
```
Right, yes

litherum reviewed (approved):
r=me assuming that `WGPUBufferBindingType_Float3x3` and `WGPUBufferBindingType_Float4x3` will go away once we get rid of our reflection code. It has to be valid to create a BGL and BG before seeing any shaders at all. Instead, the BGL itself gets to _just decide_ what the internal layout of the BGL is, and then the BG creation code, and the shader compiler, need to follow it (ideally via using a shared function that both of them call)

litherum commented on Source/WebGPU/WebGPU/Device.h:
```diff
@@ -153,6 +165,9 @@ class Device : public WGPUDeviceImpl, public ThreadSafeRefCountedAndCanMakeThrea
 
     const Ref<Adapter> m_adapter;
+#if defined(__OBJC__) && COREVIDEO_SUPPORTS_METAL
+    CVMetalTextureCacheRef m_coreVideoTextureCache;
```
I think this has to be `RetainPtr<CVMetalTextureCacheRef> m_coreVideoTextureCache;` because ARC doesn't `CFRelease()` CF types.

litherum commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -151,6 +152,9 @@ static void captureFrame(id<MTLDevice> captureObject)
 #endif
 
+#if COREVIDEO_SUPPORTS_METAL
+    CVMetalTextureCacheCreate(nullptr, nullptr, device, nullptr, &m_coreVideoTextureCache);
```
Should we check the return code?

mwyrzykowski committed:
[WebGPU] Implement GPUExternalTexture
https://bugs.webkit.org/show_bug.cgi?id=255883
<radar://108463707>

Reviewed by Myles C. Maxfield.

Implement GPUExternalTexture and pass it through to WebGPU.framework.

* Source/WebCore/Modules/WebGPU/GPUExternalTextureDescriptor.h:
(WebCore::GPUExternalTextureDescriptor::convertToBacking const):

* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUConvertToBackingContext.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
(PAL::WebGPU::DeviceImpl::importExternalTexture):
(PAL::WebGPU::DeviceImpl::createBindGroupLayout):
(PAL::WebGPU::DeviceImpl::createBindGroup):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDowncastConvertToBackingContext.cpp:
(PAL::WebGPU::DowncastConvertToBackingContext::convertToBacking):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDowncastConvertToBackingContext.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUExternalTextureImpl.cpp:
(PAL::WebGPU::ExternalTextureImpl::ExternalTextureImpl):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUExternalTextureImpl.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUExternalTextureDescriptor.h:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/APIConversions.h:
(WebGPU::fromAPI):
* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::Device::createBindGroup):
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::Device):
(WebGPU::metalPixelFormatForCVPixelBuffer):
(WebGPU::Device::createTextureFromPixelBuffer const):
(wgpuDeviceCreateExternalTexture):
* Source/WebGPU/WebGPU/ExternalTexture.h: Added.
(WebGPU::ExternalTexture::create):
(WebGPU::ExternalTexture::createInvalid):
(WebGPU::ExternalTexture::pixelBuffer const):
* Source/WebGPU/WebGPU/ExternalTexture.mm: Added.
(WebGPU::Device::createExternalTexture):
(WebGPU::ExternalTexture::ExternalTexture):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::createRemoteGPU):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteAdapter.cpp:
(WebKit::RemoteAdapter::RemoteAdapter):
(WebKit::RemoteAdapter::requestDevice):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteAdapter.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
(WebKit::RemoteDevice::RemoteDevice):
(WebKit::RemoteDevice::importExternalTexture):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
(WebKit::RemoteGPU::RemoteGPU):
(WebKit::RemoteGPU::remoteRenderingBackend const):
(WebKit::RemoteGPU::requestAdapter):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderPassEncoder.cpp:
(WebKit::RemoteRenderPassEncoder::setBindGroup):
* Source/WebKit/Shared/WebGPU/WebGPUExternalTextureDescriptor.cpp:
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/WebGPU/WebGPUExternalTextureDescriptor.h:
* Source/WebKit/Shared/WebGPU/WebGPUExternalTextureDescriptor.serialization.in:
* Websites/webkit.org/demos/webgpu/external-textured-cube.html: Added.
* Websites/webkit.org/demos/webgpu/scripts/external-textured-cube.js: Added.
(async helloCube.frameUpdate):
(async helloCube):

Canonical link: https://commits.webkit.org/264108@main
